services/src/main/java/org/keycloak/truststore/TruststoreBuilder.java

```diff
@@ -0,0 +1,221 @@
+    private static final String CERT_PROTECTION_ALGORITHM_KEY = "keystore.pkcs12.certProtectionAlgorithm";
+    static final String DUMMY_PASSWORD = "keycloakchangeit"; // fips length compliant dummy password
+    static final String PKCS12 = "PKCS12";
+
+    private static final Logger LOGGER = Logger.getLogger(TruststoreBuilder.class);
+
+    public static void setSystemTruststore(String[] truststores, boolean trustStoreIncludeDefault, String dataDir) {
+        KeyStore truststore = createMergedTruststore(truststores, trustStoreIncludeDefault);
+
+        // save with a dummy password just in case some logic that uses the system properties needs to have one
+        File file = saveTruststore(truststore, dataDir, DUMMY_PASSWORD.toCharArray());
+
+        // finally update the system properties
+        System.setProperty(TruststoreBuilder.SYSTEM_TRUSTSTORE_KEY, file.getAbsolutePath());
+        System.setProperty(TruststoreBuilder.SYSTEM_TRUSTSTORE_TYPE_KEY, PKCS12);
+        System.setProperty(TruststoreBuilder.SYSTEM_TRUSTSTORE_PASSWORD_KEY, DUMMY_PASSWORD);
+    }
+
+    static File saveTruststore(KeyStore truststore, String dataDir, char[] password) {
+        File file = new File(dataDir, "keycloak-truststore.p12");
+        file.getParentFile().mkdirs();
+        try (FileOutputStream fos = new FileOutputStream(file)) {
+            // this should inhibit the use of encryption in storing the certs
+            // it's of course not concurrency safe, but it should only be run at startup
+            String oldValue = System.setProperty(CERT_PROTECTION_ALGORITHM_KEY, "NONE");
+            truststore.store(fos, password);
+            if (oldValue != null) {
+                System.setProperty(CERT_PROTECTION_ALGORITHM_KEY, oldValue);
+            } else {
+                System.getProperties().remove(CERT_PROTECTION_ALGORITHM_KEY);
+            }
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to save truststore: " + file.getAbsolutePath(), e);
+        }
+        return file;
+    }
+
+    static KeyStore createMergedTruststore(String[] truststores, boolean trustStoreIncludeDefault) {
+        KeyStore truststore = createPkcs12KeyStore();
+
+        if (trustStoreIncludeDefault) {
+            includeDefaultTruststore(truststore);
+        }
+
+        mergeFiles(truststores, truststore);
+        return truststore;
+    }
+
+    private static void mergeFiles(String[] truststores, KeyStore truststore) {
+        for (String file : truststores) {
+            File f = new File(file);
+            if (f.isDirectory()) {
+                mergeFiles(Stream.of(f.listFiles()).map(File::getAbsolutePath).toArray(String[]::new), truststore);
+            } else {
+                if (file.endsWith(".p12") || file.endsWith(".pfx")) {
+                    mergeTrustStore(truststore, file, loadStore(file, PKCS12, null));
+                } else {
+                    mergePemFile(truststore, file);
```

Comment:
> But what about PEMs?

I couldn't find a definite list of pem extensions - crt seems to be the most common, but sometimes it lacks an extension.  

> Should just swallow some exceptions?

We briefly touched on that elsewhere and just left it as letting the exception happen.  It's certainly fine if you want it to log / swallow unrecognized files - the one corner case to that is a file contains an invalid cert, we'll log that instead.

Reply:
From my perspective, I'd just log/swallow the exception if the PEM file is unreadable. That'd make probably sense only when the path is a dir, and not a file directly... We should still throw an exception if a user points to a file directly and it can't be read correctly.

Reply:
The exception has been changed to an info log.